### PR TITLE
Authentication does handle resends

### PR DESCRIPTION
**Pull request by jrw972**
An auth resend can cause the authentication process to get into a bad
state (stall and/or livelock).

Solution: Rework the state machine with the following objectives
1. Use sequence numbers for the stateless endpoints for at-most-once
   processing.
2. Resend all handshake messages (request, reply, final) according to
   the resend timer.  Finals are resent until a token is received from
   the replier.
3. Resend auth reqs until a handshake message is received from the remote.
4. Further separate auth state from handshake state.  So that a
   participant remains authenticated through a reauth scenario.

### CODE REVIEW

File: dds/DCPS/RTPS/Spdp.cpp [hunk 1]
```diff
@@ -983,6 +1003,123 @@ Spdp::send_handshake_request(const DCPS::RepoId& guid, DiscoveredParticipant& dp
...
+  PendingRemoteAuthTokenMap::iterator token_iter = pending_remote_auth_tokens_.find(guid);
+  if (token_iter == pending_remote_auth_tokens_.end()) {
+    dp.remote_auth_request_token_ = DDS::Security::Token();
+  } else {
+    dp.remote_auth_request_token_ = token_iter->second;
+    pending_remote_auth_tokens_.erase(token_iter);
+  }
+
+  if (DCPS::security_debug.auth_debug) {
+    ACE_DEBUG((LM_DEBUG, "(%P|%t) {auth_debug} DEBUG: Spdp::attempt_authentication for %C have_remote_token=%d auth_state=%d handshake_state=%d\n", OPENDDS_STRING(DCPS::GuidConverter(guid)).c_str(), !(dp.remote_auth_request_token_ == DDS::Security::Token()), dp.auth_state_, dp.handshake_state_));
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:1021]: Add `from_discovery` to log

File: dds/DCPS/RTPS/Spdp.cpp [hunk 2]
```diff
@@ -983,6 +1003,123 @@ Spdp::send_handshake_request(const DCPS::RepoId& guid, DiscoveredParticipant& dp
...
+  DDS::Security::Authentication_var auth = security_config_->get_authentication();
+  DDS::Security::SecurityException se = {"", 0, 0};
+
+  const DDS::Security::ValidationResult_t vr = auth->validate_remote_identity(dp.identity_handle_,
+                                                                              dp.local_auth_request_token_, dp.remote_auth_request_token_, identity_handle_, dp.identity_token_, guid, se);
+
+  dp.have_auth_req_msg_ = !(dp.local_auth_request_token_ == DDS::Security::Token());
+  if (dp.have_auth_req_msg_) {
+    dp.auth_req_msg_.message_identity.source_guid = guid_;
+    dp.auth_req_msg_.message_identity.sequence_number = static_cast<CORBA::LongLong>((++stateless_sequence_number_).getValue());
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:1044]: Is this cast needed?  Both source and target are 64-bit signed.

File: dds/DCPS/RTPS/Spdp.cpp [hunk 3]
```diff
@@ -983,6 +1003,123 @@ Spdp::send_handshake_request(const DCPS::RepoId& guid, DiscoveredParticipant& dp
...
+    dp.auth_req_msg_.related_message_identity.source_guid = GUID_UNKNOWN;
+    dp.auth_req_msg_.related_message_identity.sequence_number = 0;
+    dp.auth_req_msg_.message_data.length(1);
+    dp.auth_req_msg_.message_data[0] = dp.local_auth_request_token_;
+    // Send the auth req immediately to reset the remote if they are
+    // still authenticated with us.
+    if (sedp_.write_stateless_message(dp.auth_req_msg_, make_id(guid, ENTITYID_P2P_BUILTIN_PARTICIPANT_STATELESS_READER)) != DDS::RETCODE_OK) {
+      if (DCPS::security_debug.auth_debug) {
+        ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) {auth_debug} Spdp::attempt_authentication() - ")
+                   ACE_TEXT("Unable to write auth req message retry.\n")));
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:1058]:
```suggestion
                   ACE_TEXT("Unable to write auth req message.\n")));
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 4]
```diff
@@ -983,6 +1003,123 @@ Spdp::send_handshake_request(const DCPS::RepoId& guid, DiscoveredParticipant& dp
...
+      }
+    } else {
+      if (DCPS::security_debug.auth_debug) {
+        ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) {auth_debug} DEBUG: Spdp::attempt_authentication() - ")
+                   ACE_TEXT("Sent auth req message for participant: %C\n"),
+                   OPENDDS_STRING(DCPS::GuidConverter(guid)).c_str()));
+      }
+    }
+    // Schedule resend.
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:1067]:
```suggestion
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 5]
```diff
@@ -983,6 +1003,123 @@ Spdp::send_handshake_request(const DCPS::RepoId& guid, DiscoveredParticipant& dp
...
+    schedule_handshake_resend(config_->auth_resend_period(), guid);
+  }
+
+  switch (vr) {
+  case DDS::Security::VALIDATION_OK: {
+    dp.auth_state_ = DCPS::AUTH_STATE_AUTHENTICATED;
+    dp.handshake_state_ = DCPS::HANDSHAKE_STATE_DONE;
+    purge_handshake_deadlines(participants_.find(guid));
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:1075]:
```suggestion
    purge_handshake_deadlines(iter);
```
I think there are a few of these

File: dds/DCPS/RTPS/Spdp.cpp [hunk 6]
```diff
@@ -1012,248 +1149,234 @@ Spdp::handle_handshake_message(const DDS::Security::ParticipantStatelessMessage&
...
-                   ACE_TEXT("Dropping handshake message from %C, incorrect related message identity source guid\n"),
-                   OPENDDS_STRING(DCPS::GuidConverter(src_participant)).c_str()));
-      }
-      return;
-    }
+  case DCPS::HANDSHAKE_STATE_BEGIN_HANDSHAKE_REQUEST:
+  case DCPS::HANDSHAKE_STATE_WAITING_FOR_TOKEN:
+  case DCPS::HANDSHAKE_STATE_DONE: {
+    ACE_ERROR((LM_ERROR,
+               ACE_TEXT("(%P|%t) {auth_warn} Spdp::handle_handshake_message() - ")
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:1165]:
```suggestion
               ACE_TEXT("(%P|%t) Spdp::handle_handshake_message() - ")
```

File: dds/DCPS/RTPS/Spdp.cpp [hunk 7]
```diff
@@ -1012,248 +1149,234 @@ Spdp::handle_handshake_message(const DDS::Security::ParticipantStatelessMessage&
...
           ACE_DEBUG((LM_WARNING, ACE_TEXT("(%P|%t) {auth_warn} WARNING: Spdp::handle_handshake_message() - ")
-                     ACE_TEXT("Unexpected validation pending retry\n")));
+                     ACE_TEXT("Unable to write stateless message for handshake reply.\n")));
         }
         return;
-      case DDS::Security::VALIDATION_PENDING_HANDSHAKE_REQUEST:
+      } else {
+        if (DCPS::security_debug.auth_debug) {
+          ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) {auth_debug} DEBUG: Spdp::handle_handshake_message() - ")
+                     ACE_TEXT("Sent hanshake unknown message for participant: %C\n"),
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Spdp.cpp:1327]:
```suggestion
                     ACE_TEXT("Sent handshake unknown message for participant: %C\n"),
```

File: dds/DCPS/RTPS/Spdp.cpp (hunk 6 above)
Comment (iguessthislldo) [reply dds/DCPS/RTPS/Spdp.cpp:1165]: Or make it actually make it conditional on the `auth_warn` logging tag.

**Pull request merged**